extensions/amp-next-page/0.2/visibility-observer.js

```diff
@@ -0,0 +1,190 @@
+      .getPositionObserver()
+      .observe(bottom, PositionObserverFidelity.LOW, position =>
+        this.bottomSentinelPositionChanged(position, callback)
+      );
+  }
+
+  /**
+   * Called when a position change is detected on the injected
+   * top sentinel element
+   * @param {?PositionInViewportEntryDef} position
+   * @param {!Function} callback
+   */
+  topSentinelPositionChanged(position, callback) {
+    const prevTopSentinelPosition = this.topSentinelPosition_;
+    if (position.relativePos === prevTopSentinelPosition) {
+      return;
+    }
+    this.topSentinelPosition_ = position.relativePos;
+    this.updateRelativePos_(callback);
+  }
+
+  /**
+   * Called when a position change is detected on the injected
+   * bottom sentinel element
+   * @param {?PositionInViewportEntryDef} position
+   * @param {!Function} callback
+   */
+  bottomSentinelPositionChanged(position, callback) {
+    const prevBottomSentinelPosition = this.bottomSentinelPosition_;
+    if (position.relativePos === prevBottomSentinelPosition) {
+      return;
+    }
+    this.bottomSentinelPosition_ = position.relativePos;
+    this.updateRelativePos_(callback);
+  }
+
+  /**
+   * Calculates the position of the element relative to the viewport
+   * based on the positions of the injected bottom and top sentinel elements
+   * @param {!Function} callback
+   * @private
+   */
+  updateRelativePos_(callback) {
+    const {topSentinelPosition_: top, bottomSentinelPosition_: bottom} = this;
+    const {INSIDE, TOP, BOTTOM} = RelativePositions;
+
+    devAssert(
+      top || bottom,
+      'scroll triggered without a top or bottom sentinel position'
+    );
+
+    if (top === INSIDE && bottom === INSIDE) {
+      // Both the top and bottom sentinel elements are within the
+      // viewport bounds meaning that the document is short enough
+      // to be contained inside the viewport
+      this.relativePos_ = ViewportRelativePos.INSIDE_VIEWPORT;
+    } else if ((!top || top === TOP) && (!bottom || bottom === BOTTOM)) {
+      // The head of the document is above the viewport and the
+      // foot of the document is below it, meaning that the viewport
+      // is looking at a section of the document
+      this.relativePos_ = ViewportRelativePos.CONTAINS_VIEWPORT;
+    } else if (
+      ((!top || top === TOP) && bottom === TOP) ||
+      (top === BOTTOM && (!bottom || bottom === BOTTOM))
+    ) {
+      // Both the top and the bottom of the document are either
+      // above or below the document meaning that the viewport hasn't
+      // reached the document yet or has passed it
+      this.relativePos_ = ViewportRelativePos.OUTSIDE_VIEWPORT;
+    } else {
+      // The remaining case is the case where the document is halfway
+      // through being scrolling into/out of the viewport in which case
+      // we don't need to update the visibility
+      this.relativePos_ = ViewportRelativePos.LEAVING_VIEWPORT;
+      return;
+    }
+
+    if (callback) {
```

Comment:
The function parameter here is non-nullable, and all the chain of calls also has it non-nullable, so you shouldn't need an if-check here.